pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/table/internal/JavaResolvers.java

```diff
@@ -73,6 +74,39 @@ static boolean canBeImportedIn(String thisPackage, JAccessibleElementSymbol memb
         return isAccessibleIn(null, thisPackage, member, false);
     }
 
+    @NonNull
+    static NameResolver<JTypeMirror> moduleImport(Set<String> moduleNames,
+                                                  final SymbolResolver symbolResolver,
+                                                  final String thisPackage) {
+        return new SingleNameResolver<JTypeMirror>() {
+            @Override
+            public @Nullable JTypeMirror resolveFirst(String simpleName) {
+                for (String module : moduleNames) {
+                    JModuleSymbol moduleSymbol = symbolResolver.resolveModule(module);
+
+                    if (moduleSymbol == null) {
+                        // TODO: log warning about incomplete auxclasspath
+                        return null;
+                    }
+
+                    for (String packageName : moduleSymbol.getExportedPackages()) {
+                        JClassSymbol sym = symbolResolver.resolveClassFromCanonicalName(packageName + "." + simpleName);
```

Comment:
This should resolve from binary name instead of canonical name, otherwise it will try every package segment. But we know that these segments are packages and not outer classes. This is different from type import on demand, because these can import inner types too.